language/tools/test-generation/src/main.rs

```diff
@@ -3,10 +3,67 @@
 
 #![forbid(unsafe_code)]
 
+use num_cpus;
+use rayon::prelude::*;
+use slog::{o, Drain};
+use std::{env, ops::Range};
 use structopt::StructOpt;
 use test_generation::{config::Args, run_generation};
 
+fn partition_work(args: &Args) -> Vec<Option<Range<u64>>> {
+    let num_worker_threads = if let Some(num_threads) = args.num_threads {
+        num_threads
+    } else {
+        num_cpus::get() as u64
+    };
+    assert!(
+        num_worker_threads > 0,
+        "Number of worker threads must be greater than 0"
+    );
+    assert!(
+        args.num_iterations.is_some(),
+        "TODO: number of iterations must be supplied for now"
+    );
+    match args.num_iterations {
+        None => (0..num_worker_threads).map(|_| None).collect(),
+        Some(num_iters) => {
+            let div = num_iters / num_worker_threads;
+            let rem = num_iters % num_worker_threads;
+            // Distribute the remainder workload evenly
+            let mut acc = Vec::new();
+            let mut start_index = 0;
+            for i in 0..num_worker_threads {
+                let end_index = if i >= num_worker_threads - rem {
+                    start_index + div + 1
+                } else {
+                    start_index + div
+                };
+                acc.push(Some(start_index..end_index));
+                start_index = end_index;
+            }
+            acc
+        }
+    }
+}
+
+fn setup_log() {
+    if env::var("RUST_LOG").is_err() {
+        env::set_var("RUST_LOG", "info");
+    }
+    let decorator = slog_term::PlainDecorator::new(std::io::stdout());
+    let drain = slog_term::CompactFormat::new(decorator).build().fuse();
+    let drain = slog_envlogger::new(drain);
+    let drain = std::sync::Mutex::new(drain).fuse();
+    let logger = slog::Logger::root(drain, o!());
+    let logger_guard = slog_scope::set_global_logger(logger);
+    std::mem::forget(logger_guard);
+}
+
 pub fn main() {
+    setup_log();
     let args = Args::from_args();
-    run_generation(args);
+    let workload = partition_work(&args);
+    workload.into_par_iter().for_each(|work_segment| {
```

Comment:
What does a work_segment represent?

Reply:
A work segment represents the amount of work for a single thread to perform. The general gist is we partition the number of tests to generate equally amongst all available threads at the beginning (into `work_segments`), we then assign each work segment to a thread using this `par_iter`.

Reply:
Just FYI: This concept no longer exists with the new parallelization strategy.